### ISSUE TITLE
Improve priorities on overlapping rules

### ISSUE THREAD

**Issue by jaumeortola** [languagetool-org/languagetool#2780]
When there are overlapping rules, it is essential to choose the best suited to the context so that the users are not confused. 

The current criteria to choose between overlapping rules are a bit simple. The first appearing and the longest suggestions are always chosen. We need to make these criteria a bit more sophisticated. 

Related to #2778.

**Comment by jaumeortola** [comment]
Example (spanish): "Esto esta rico".

The best suggestion is "esta > está", but it is shorter than "esta rico> este rico" and it is not shown.

**Comment by tiff** [comment]
@jaumeortola I don't know if that helps but we have something like that for English and German already. Check English.java

**Comment by jaumeortola** [comment]
@tiff Do you mean "getPriorityForId" or some other thing? This is already used in most languages. But it cannot fix the problem.

**Comment by tiff** [comment]
Yeah, I mean `getPriorityForId`. Sorry, I might have misunderstood the issue you are having.

**Comment by jaumeortola** [comment]
Indeed, `getPriorityForId` solves most problems (including my previous example). I didn't remember well all the criteria. 

Sometimes it would be necessary to apply `getPriorityForId` even in juxtaposed errors, which isn't done now. See: https://github.com/languagetool-org/languagetool/blob/master/languagetool-core/src/main/java/org/languagetool/rules/CleanOverlappingFilter.java#L5. But changing this could create other undesired results. I will try to do it by other means. 

Example of juxtaposed errors (Spanish): "el termino > él termino / el término / él terminó".
